File: artifacts/api-server/src/repositories/postgres/timesheet-repository.ts
import { db } from "@workspace/db";
import { timesheetsTable, timesheetRowsTable, employeesTable, projectsTable, activitiesTable } from "@workspace/db";
import { eq, and, sql, inArray } from "drizzle-orm";
import { generateId } from "../../lib/id.js";
import type { ITimesheetRepository } from "../interfaces.js";

function calcTotal(row: any) {
  return (row.monday || 0) + (row.tuesday || 0) + (row.wednesday || 0) + (row.thursday || 0) + (row.friday || 0) + (row.saturday || 0) + (row.sunday || 0);
}

function weekEndDate(weekStartDate: string) {
  const d = new Date(weekStartDate);
  d.setDate(d.getDate() + 6);
  return d.toISOString().split("T")[0];
}

async function enrichTimesheet(ts: any) {
  const rows = await db.select().from(timesheetRowsTable).where(eq(timesheetRowsTable.timesheetId, ts.id));
  const employee = await db.select().from(employeesTable).where(eq(employeesTable.id, ts.employeeId));
  let approverName = null;
  if (ts.approvedBy) {
    const approver = await db.select().from(employeesTable).where(eq(employeesTable.id, ts.approvedBy));
    approverName = approver[0]?.name ?? null;
  }
  const enrichedRows = await Promise.all(rows.map(async (r) => {
    const project = await db.select().from(projectsTable).where(eq(projectsTable.id, r.projectId));
    const activity = await db.select().from(activitiesTable).where(eq(activitiesTable.id, r.activityId));
    return { ...r, projectName: project[0]?.name ?? null, activityName: activity[0]?.name ?? null };
  }));
  return {
    ...ts,
    employeeName: employee[0]?.name ?? null,
    approverName,
    rows: enrichedRows,
  };
}

export class PostgresTimesheetRepository implements ITimesheetRepository {
  async findAll(params: { page?: number; pageSize?: number; employeeId?: string; status?: string; weekStartDate?: string; managerId?: string }) {
    const page = params.page ?? 1;
    const pageSize = params.pageSize ?? 20;
    const offset = (page - 1) * pageSize;

    let baseQuery = `SELECT t.* FROM timesheets t`;
    const conditions: string[] = [];
    const values: any[] = [];
    let idx = 1;

    if (params.managerId) {
      baseQuery += ` JOIN employees e ON t.employee_id = e.id`;
      conditions.push(`e.manager_id = $${idx++}`);
      values.push(params.managerId);
    }
    if (params.employeeId) { conditions.push(`t.employee_id = $${idx++}`); values.push(params.employeeId); }
    if (params.status) { conditions.push(`t.status = $${idx++}`); values.push(params.status); }
    if (params.weekStartDate) { conditions.push(`t.week_start_date = $${idx++}`); values.push(params.weekStartDate); }

    const whereClause = conditions.length > 0 ? ` WHERE ${conditions.join(" AND ")}` : "";
    const countResult = await db.execute(sql.raw(`SELECT COUNT(*) as count FROM (${baseQuery}${whereClause}) sub`, values));
    const rowsResult = await db.execute(sql.raw(`${baseQuery}${whereClause} ORDER BY t.week_start_date DESC, t.created_at DESC LIMIT ${pageSize} OFFSET ${offset}`, values));

    const rows = rowsResult.rows as any[];
    const total = Number((countResult.rows as any[])[0]?.count || 0);
    const enriched = await Promise.all(rows.map(enrichTimesheet));
    return { data: enriched, total, page, pageSize };
  }

  async findById(id: string) {
    const rows = await db.select().from(timesheetsTable).where(eq(timesheetsTable.id, id));
    if (!rows[0]) return null;
    return enrichTimesheet(rows[0]);
  }

  async findByEmployeeAndWeek(employeeId: string, weekStartDate: string) {
    const rows = await db.select().from(timesheetsTable).where(
      and(eq(timesheetsTable.employeeId, employeeId), eq(timesheetsTable.weekStartDate, weekStartDate))
    );
    if (!rows[0]) return null;
    return enrichTimesheet(rows[0]);
  }

  async create(data: any, rows: any[]) {
    const id = generateId();
    const now = new Date();
    const endDate = weekEndDate(data.weekStartDate);
    let totalHours = 0;

    await db.insert(timesheetsTable).values({
      id, employeeId: data.employeeId, weekStartDate: data.weekStartDate,
      weekEndDate: endDate, status: "Draft", totalHours: 0, createdAt: now, updatedAt: now,
    });

    for (const row of rows) {
      const rowId = generateId();
      const total = calcTotal(row);
      totalHours += total;
      await db.insert(timesheetRowsTable).values({
        id: rowId, timesheetId: id, projectId: row.projectId, activityId: row.activityId,
        monday: row.monday || 0, tuesday: row.tuesday || 0, wednesday: row.wednesday || 0,
        thursday: row.thursday || 0, friday: row.friday || 0, saturday: row.saturday || 0,
        sunday: row.sunday || 0, totalHours: total, comments: row.comments ?? null, createdAt: now,
      });
    }

    await db.update(timesheetsTable).set({ totalHours, updatedAt: now }).where(eq(timesheetsTable.id, id));
    return this.findById(id);
  }

  async update(id: string, rows: any[]) {
    const ts = await db.select().from(timesheetsTable).where(eq(timesheetsTable.id, id));
    if (!ts[0] || ts[0].status !== "Draft") return null;
    await db.delete(timesheetRowsTable).where(eq(timesheetRowsTable.timesheetId, id));
    const now = new Date();
    let totalHours = 0;
    for (const row of rows) {
      const rowId = generateId();
      const total = calcTotal(row);
      totalHours += total;
      await db.insert(timesheetRowsTable).values({
        id: rowId, timesheetId: id, projectId: row.projectId, activityId: row.activityId,
        monday: row.monday || 0, tuesday: row.tuesday || 0, wednesday: row.wednesday || 0,
        thursday: row.thursday || 0, friday: row.friday || 0, saturday: row.saturday || 0,
        sunday: row.sunday || 0, totalHours: total, comments: row.comments ?? null, createdAt: now,
      });
    }
    await db.update(timesheetsTable).set({ totalHours, updatedAt: now }).where(eq(timesheetsTable.id, id));
    return this.findById(id);
  }

  async submit(id: string) {
    await db.update(timesheetsTable).set({ status: "Submitted", submittedAt: new Date(), updatedAt: new Date() }).where(eq(timesheetsTable.id, id));
    return this.findById(id);
  }

  async approve(id: string, approvedBy: string, comment?: string) {
    await db.update(timesheetsTable).set({ status: "Approved", approvedAt: new Date(), approvedBy, updatedAt: new Date() }).where(eq(timesheetsTable.id, id));
    return this.findById(id);
  }

  async reject(id: string, comment: string) {
    await db.update(timesheetsTable).set({ status: "Rejected", rejectionComment: comment, updatedAt: new Date() }).where(eq(timesheetsTable.id, id));
    return this.findById(id);
  }

  async bulkAction(ids: string[], action: "approve" | "reject", approvedBy: string, comment?: string) {
    let succeeded = 0, failed = 0;
    for (const id of ids) {
      try {
        if (action === "approve") await this.approve(id, approvedBy, comment);
        else await this.reject(id, comment ?? "Rejected");
        succeeded++;
      } catch { failed++; }
    }
    return { processed: ids.length, succeeded, failed };
  }

  async copyFromPreviousWeek(employeeId: string, sourceWeekStartDate: string, targetWeekStartDate: string) {
    const source = await this.findByEmployeeAndWeek(employeeId, sourceWeekStartDate);
    if (!source || !source.rows) throw new Error("Source week not found");
    const existing = await this.findByEmployeeAndWeek(employeeId, targetWeekStartDate);
    if (existing) return existing;
    const rows = source.rows.map((r: any) => ({
      projectId: r.projectId, activityId: r.activityId,
      monday: r.monday, tuesday: r.tuesday, wednesday: r.wednesday,
      thursday: r.thursday, friday: r.friday, saturday: r.saturday, sunday: r.sunday,
      comments: r.comments,
    }));
    return this.create({ employeeId, weekStartDate: targetWeekStartDate }, rows);
  }

  async getStatusBreakdown() {
    const result = await db.execute(sql`
      SELECT
        COUNT(CASE WHEN status = 'Draft' THEN 1 END) as draft,
        COUNT(CASE WHEN status = 'Submitted' THEN 1 END) as submitted,
        COUNT(CASE WHEN status = 'Approved' THEN 1 END) as approved,
        COUNT(CASE WHEN status = 'Rejected' THEN 1 END) as rejected
      FROM timesheets
    `);
    const r = (result.rows as any[])[0] || {};
    return { draft: Number(r.draft||0), submitted: Number(r.submitted||0), approved: Number(r.approved||0), rejected: Number(r.rejected||0) };
  }

  async getRecentActivity(limit: number) {
    const result = await db.execute(sql`
      SELECT t.id as timesheet_id, t.status, t.week_start_date, t.updated_at as timestamp, e.name as employee_name
      FROM timesheets t
      JOIN employees e ON t.employee_id = e.id
      ORDER BY t.updated_at DESC
      LIMIT ${limit}
    `);
    return (result.rows as any[]).map((r) => ({
      id: r.timesheet_id,
      action: r.status,
      employeeName: r.employee_name,
      weekStartDate: r.week_start_date,
      timestamp: r.timestamp,
      timesheetId: r.timesheet_id,
    }));
  }

  async getComplianceOverview() {
    const result = await db.execute(sql`
      SELECT 
        e.department,
        COUNT(DISTINCT e.id) as total_employees,
        COUNT(DISTINCT CASE WHEN t.status IN ('Submitted','Approved') THEN e.id END) as submitted
      FROM employees e
      LEFT JOIN timesheets t ON e.id = t.employee_id AND t.week_start_date >= (CURRENT_DATE - INTERVAL '4 weeks')::text
      WHERE e.status = 'Active'
      GROUP BY e.department
      ORDER BY e.department
    `);
    return (result.rows as any[]).map((r) => ({
      department: r.department,
      totalEmployees: Number(r.total_employees),
      submitted: Number(r.submitted),
      complianceRate: Number(r.total_employees) > 0 ? Math.round((Number(r.submitted) / Number(r.total_employees)) * 100) : 0,
    }));
  }
}

File: artifacts/api-server/Repositories/TimesheetRepository.cs
using Microsoft.EntityFrameworkCore;
using Npgsql;
using TimesheetApi.Data;
using TimesheetApi.Helpers;
using TimesheetApi.Models;

namespace TimesheetApi.Repositories;

public class TimesheetRepository : ITimesheetRepository
{
    private readonly AppDbContext _db;
    private readonly string _connectionString;

    public TimesheetRepository(AppDbContext db, IConfiguration configuration)
    {
        _db = db;
        _connectionString = configuration["ConnectionStrings:Default"]
            ?? Environment.GetEnvironmentVariable("DATABASE_URL")
            ?? throw new InvalidOperationException("DATABASE_URL not set");
    }

    public async Task<PaginatedResult<object>> FindAll(int page, int pageSize, string? employeeId,
        string? status, string? weekStartDate, string? managerId)
    {
        var offset = (page - 1) * pageSize;
        var conditions = new List<string>();
        var values = new List<object>();
        int idx = 1;

        var baseQuery = "SELECT t.* FROM timesheets t";

        if (!string.IsNullOrEmpty(managerId))
        {
            baseQuery += " JOIN employees e ON t.employee_id = e.id";
            conditions.Add($"e.manager_id = ${idx++}");
            values.Add(managerId);
        }
        if (!string.IsNullOrEmpty(employeeId)) { conditions.Add($"t.employee_id = ${idx++}"); values.Add(employeeId); }
        if (!string.IsNullOrEmpty(status)) { conditions.Add($"t.status::text = ${idx++}"); values.Add(status); }
        if (!string.IsNullOrEmpty(weekStartDate)) { conditions.Add($"t.week_start_date = ${idx++}"); values.Add(weekStartDate); }

        var where = conditions.Count > 0 ? $" WHERE {string.Join(" AND ", conditions)}" : "";
        var valArr = values.ToArray();

        int total;
        var rawRows = new List<Dictionary<string, object?>>();

        await using var conn = new NpgsqlConnection(_connectionString);
        await conn.OpenAsync();

        await using (var countCmd = new NpgsqlCommand($"SELECT COUNT(*) FROM ({baseQuery}{where}) sub", conn))
        {
            for (int i = 0; i < valArr.Length; i++) countCmd.Parameters.AddWithValue(valArr[i]);
            total = Convert.ToInt32(await countCmd.ExecuteScalarAsync());
        }

        await using (var rowCmd = new NpgsqlCommand(
            $"{baseQuery}{where} ORDER BY t.week_start_date DESC, t.created_at DESC LIMIT {pageSize} OFFSET {offset}", conn))
        {
            for (int i = 0; i < valArr.Length; i++) rowCmd.Parameters.AddWithValue(valArr[i]);
            await using var reader = await rowCmd.ExecuteReaderAsync();
            while (await reader.ReadAsync())
                rawRows.Add(ReadTimesheetRow(reader));
        }

        var enriched = await Task.WhenAll(rawRows.Select(r => EnrichTimesheet(r)));
        return new PaginatedResult<object>(enriched.ToList<object>(), total, page, pageSize);
    }

    public async Task<object?> FindById(string id)
    {
        var ts = await _db.Timesheets.FirstOrDefaultAsync(t => t.Id == id);
        if (ts == null) return null;
        return await EnrichTimesheet(TimesheetToDict(ts));
    }

    public async Task<object?> FindByEmployeeAndWeek(string employeeId, string weekStartDate)
    {
        var ts = await _db.Timesheets.FirstOrDefaultAsync(t =>
            t.EmployeeId == employeeId && t.WeekStartDate == weekStartDate);
        if (ts == null) return null;
        return await EnrichTimesheet(TimesheetToDict(ts));
    }

    public async Task<object?> Create(string employeeId, string weekStartDate, List<Dictionary<string, object?>> rows)
    {
        var id = IdGenerator.NewId();
        var now = DateTime.UtcNow;
        var endDate = WeekEndDate(weekStartDate);
        float totalHours = 0;

        _db.Timesheets.Add(new Timesheet
        {
            Id = id, EmployeeId = employeeId, WeekStartDate = weekStartDate,
            WeekEndDate = endDate, Status = "Draft", TotalHours = 0,
            CreatedAt = now, UpdatedAt = now,
        });
        await _db.SaveChangesAsync();

        foreach (var row in rows)
        {
            var total = CalcTotal(row);
            totalHours += total;
            _db.TimesheetRows.Add(BuildRow(IdGenerator.NewId(), id, row, total, now));
        }

        var ts = await _db.Timesheets.FindAsync(id);
        ts!.TotalHours = totalHours;
        ts.UpdatedAt = now;
        await _db.SaveChangesAsync();
        return await FindById(id);
    }

    public async Task<object?> Update(string id, List<Dictionary<string, object?>> rows)
    {
        var ts = await _db.Timesheets.FindAsync(id);
        if (ts == null || (ts.Status != "Draft" && ts.Status != "Rejected")) return null;

        await _db.TimesheetRows.Where(r => r.TimesheetId == id).ExecuteDeleteAsync();

        var now = DateTime.UtcNow;
        float totalHours = 0;
        foreach (var row in rows)
        {
            var total = CalcTotal(row);
            totalHours += total;
            _db.TimesheetRows.Add(BuildRow(IdGenerator.NewId(), id, row, total, now));
        }
        ts.TotalHours = totalHours;
        ts.UpdatedAt = now;
        await _db.SaveChangesAsync();
        return await FindById(id);
    }

    public async Task<object?> Submit(string id)
    {
        await _db.Timesheets.Where(t => t.Id == id)
            .ExecuteUpdateAsync(s => s
                .SetProperty(t => t.Status, "Submitted")
                .SetProperty(t => t.SubmittedAt, DateTime.UtcNow)
                .SetProperty(t => t.UpdatedAt, DateTime.UtcNow));
        return await FindById(id);
    }

    public async Task<object?> Approve(string id, string approvedBy, string? comment)
    {
        await _db.Timesheets.Where(t => t.Id == id)
            .ExecuteUpdateAsync(s => s
                .SetProperty(t => t.Status, "Approved")
                .SetProperty(t => t.ApprovedAt, DateTime.UtcNow)
                .SetProperty(t => t.ApprovedBy, approvedBy)
                .SetProperty(t => t.UpdatedAt, DateTime.UtcNow));
        return await FindById(id);
    }

    public async Task<object?> Reject(string id, string comment)
    {
        await _db.Timesheets.Where(t => t.Id == id)
            .ExecuteUpdateAsync(s => s
                .SetProperty(t => t.Status, "Rejected")
                .SetProperty(t => t.RejectionComment, comment)
                .SetProperty(t => t.UpdatedAt, DateTime.UtcNow));
        return await FindById(id);
    }

    public async Task<(int Processed, int Succeeded, int Failed)> BulkAction(
        List<string> ids, string action, string approvedBy, string? comment)
    {
        int succeeded = 0, failed = 0;
        foreach (var id in ids)
        {
            try
            {
                if (action == "approve") await Approve(id, approvedBy, comment);
                else await Reject(id, comment ?? "Rejected");
                succeeded++;
            }
            catch { failed++; }
        }
        return (ids.Count, succeeded, failed);
    }

    public async Task<object?> CopyFromPreviousWeek(string employeeId, string sourceWeekStartDate, string targetWeekStartDate)
    {
        var source = await FindByEmployeeAndWeek(employeeId, sourceWeekStartDate);
        if (source == null) throw new Exception("Source week not found");

        var existing = await FindByEmployeeAndWeek(employeeId, targetWeekStartDate);
        if (existing != null) return existing;

        dynamic src = source;
        var srcRows = src.rows as IEnumerable<object> ?? Enumerable.Empty<object>();
        var rows = new List<Dictionary<string, object?>>();
        foreach (dynamic r in srcRows)
        {
            rows.Add(new Dictionary<string, object?>
            {
                ["projectId"] = r.projectId,
                ["activityId"] = r.activityId,
                ["monday"] = r.monday,
                ["tuesday"] = r.tuesday,
                ["wednesday"] = r.wednesday,
                ["thursday"] = r.thursday,
                ["friday"] = r.friday,
                ["saturday"] = r.saturday,
                ["sunday"] = r.sunday,
                ["comments"] = r.comments,
                ["mondayStart"] = r.mondayStart,
                ["mondayEnd"] = r.mondayEnd,
                ["tuesdayStart"] = r.tuesdayStart,
                ["tuesdayEnd"] = r.tuesdayEnd,
                ["wednesdayStart"] = r.wednesdayStart,
                ["wednesdayEnd"] = r.wednesdayEnd,
                ["thursdayStart"] = r.thursdayStart,
                ["thursdayEnd"] = r.thursdayEnd,
                ["fridayStart"] = r.fridayStart,
                ["fridayEnd"] = r.fridayEnd,
                ["saturdayStart"] = r.saturdayStart,
                ["saturdayEnd"] = r.saturdayEnd,
                ["sundayStart"] = r.sundayStart,
                ["sundayEnd"] = r.sundayEnd,
            });
        }
        return await Create(employeeId, targetWeekStartDate, rows);
    }

    public async Task<object> GetStatusBreakdown()
    {
        await using var conn = new NpgsqlConnection(_connectionString);
        await conn.OpenAsync();
        await using var cmd = new NpgsqlCommand(@"
            SELECT
                COUNT(CASE WHEN status::text = 'Draft' THEN 1 END) as draft,
                COUNT(CASE WHEN status::text = 'Submitted' THEN 1 END) as submitted,
                COUNT(CASE WHEN status::text = 'Approved' THEN 1 END) as approved,
                COUNT(CASE WHEN status::text = 'Rejected' THEN 1 END) as rejected
            FROM timesheets", conn);
        await using var reader = await cmd.ExecuteReaderAsync();
        if (await reader.ReadAsync())
        {
            return new
            {
                draft = Convert.ToInt32(reader["draft"]),
                submitted = Convert.ToInt32(reader["submitted"]),
                approved = Convert.ToInt32(reader["approved"]),
                rejected = Convert.ToInt32(reader["rejected"]),
            };
        }
        return new { draft = 0, submitted = 0, approved = 0, rejected = 0 };
    }

    public async Task<List<object>> GetRecentActivity(int limit)
    {
        await using var conn = new NpgsqlConnection(_connectionString);
        await conn.OpenAsync();
        await using var cmd = new NpgsqlCommand($@"
            SELECT t.id as timesheet_id, t.status::text, t.week_start_date, t.updated_at as timestamp, e.name as employee_name
            FROM timesheets t
            JOIN employees e ON t.employee_id = e.id
            ORDER BY t.updated_at DESC
            LIMIT {limit}", conn);
        await using var reader = await cmd.ExecuteReaderAsync();
        var results = new List<object>();
        while (await reader.ReadAsync())
        {
            results.Add(new
            {
                id = reader["timesheet_id"]?.ToString(),
                action = reader["status"]?.ToString(),
                employeeName = reader["employee_name"]?.ToString(),
                weekStartDate = reader["week_start_date"]?.ToString(),
                timestamp = reader["timestamp"],
                timesheetId = reader["timesheet_id"]?.ToString(),
            });
        }
        return results;
    }

    public async Task<List<object>> GetComplianceOverview()
    {
        await using var conn = new NpgsqlConnection(_connectionString);
        await conn.OpenAsync();
        await using var cmd = new NpgsqlCommand(@"
            SELECT
                e.department,
                COUNT(DISTINCT e.id) as total_employees,
                COUNT(DISTINCT CASE WHEN t.status::text IN ('Submitted','Approved') THEN e.id END) as submitted
            FROM employees e
            LEFT JOIN timesheets t ON e.id = t.employee_id
                AND t.week_start_date >= (CURRENT_DATE - INTERVAL '4 weeks')::text
            WHERE e.status::text = 'Active'
            GROUP BY e.department
            ORDER BY e.department", conn);
        await using var reader = await cmd.ExecuteReaderAsync();
        var results = new List<object>();
        while (await reader.ReadAsync())
        {
            var total = Convert.ToInt32(reader["total_employees"]);
            var submitted = Convert.ToInt32(reader["submitted"]);
            results.Add(new
            {
                department = reader["department"]?.ToString(),
                totalEmployees = total,
                submitted,
                complianceRate = total > 0 ? (int)Math.Round((double)submitted / total * 100) : 0,
            });
        }
        return results;
    }

    private async Task<object> EnrichTimesheet(Dictionary<string, object?> ts)
    {
        var tsId = ts["id"]?.ToString() ?? "";
        var employeeId = ts["employeeId"]?.ToString() ?? "";
        var approvedBy = ts["approvedBy"]?.ToString();

        var rows = await _db.TimesheetRows.Where(r => r.TimesheetId == tsId).ToListAsync();
        var employee = await _db.Employees.FirstOrDefaultAsync(e => e.Id == employeeId);

        string? approverName = null;
        if (!string.IsNullOrEmpty(approvedBy))
        {
            var approver = await _db.Employees.FirstOrDefaultAsync(e => e.Id == approvedBy);
            approverName = approver?.Name;
        }

        var enrichedRows = await Task.WhenAll(rows.Select(async r =>
        {
            var project = await _db.Projects.FirstOrDefaultAsync(p => p.Id == r.ProjectId);
            var activity = await _db.Activities.FirstOrDefaultAsync(a => a.Id == r.ActivityId);
            return (object)new
            {
                id = r.Id,
                timesheetId = r.TimesheetId,
                projectId = r.ProjectId,
                projectName = project?.Name,
                activityId = r.ActivityId,
                activityName = activity?.Name,
                monday = r.Monday,
                tuesday = r.Tuesday,
                wednesday = r.Wednesday,
                thursday = r.Thursday,
                friday = r.Friday,
                saturday = r.Saturday,
                sunday = r.Sunday,
                totalHours = r.TotalHours,
                comments = r.Comments,
                mondayStart = r.MondayStart,
                mondayEnd = r.MondayEnd,
                tuesdayStart = r.TuesdayStart,
                tuesdayEnd = r.TuesdayEnd,
                wednesdayStart = r.WednesdayStart,
                wednesdayEnd = r.WednesdayEnd,
                thursdayStart = r.ThursdayStart,
                thursdayEnd = r.ThursdayEnd,
                fridayStart = r.FridayStart,
                fridayEnd = r.FridayEnd,
                saturdayStart = r.SaturdayStart,
                saturdayEnd = r.SaturdayEnd,
                sundayStart = r.SundayStart,
                sundayEnd = r.SundayEnd,
                createdAt = r.CreatedAt,
            };
        }));

        return new
        {
            id = ts["id"],
            employeeId = ts["employeeId"],
            employeeName = employee?.Name,
            weekStartDate = ts["weekStartDate"],
            weekEndDate = ts["weekEndDate"],
            status = ts["status"],
            totalHours = ts["totalHours"],
            submittedAt = ts["submittedAt"],
            approvedAt = ts["approvedAt"],
            approvedBy = ts["approvedBy"],
            approverName,
            rejectionComment = ts["rejectionComment"],
            createdAt = ts["createdAt"],
            updatedAt = ts["updatedAt"],
            rows = enrichedRows,
        };
    }

    private static float CalcTotal(Dictionary<string, object?> row)
    {
        float Get(string key) => row.TryGetValue(key, out var v) ? Convert.ToSingle(v ?? 0f) : 0f;
        return Get("monday") + Get("tuesday") + Get("wednesday") + Get("thursday")
             + Get("friday") + Get("saturday") + Get("sunday");
    }

    private static string WeekEndDate(string weekStartDate)
    {
        var d = DateTime.Parse(weekStartDate).AddDays(6);
        return d.ToString("yyyy-MM-dd");
    }

    private static TimesheetRow BuildRow(string rowId, string timesheetId, Dictionary<string, object?> row, float total, DateTime now)
    {
        float GetF(string key) => row.TryGetValue(key, out var v) ? Convert.ToSingle(v ?? 0f) : 0f;
        string? GetS(string key) => row.TryGetValue(key, out var v) ? v?.ToString() : null;

        return new TimesheetRow
        {
            Id = rowId,
            TimesheetId = timesheetId,
            ProjectId = GetS("projectId") ?? "",
            ActivityId = GetS("activityId") ?? "",
            Monday = GetF("monday"),
            Tuesday = GetF("tuesday"),
            Wednesday = GetF("wednesday"),
            Thursday = GetF("thursday"),
            Friday = GetF("friday"),
            Saturday = GetF("saturday"),
            Sunday = GetF("sunday"),
            TotalHours = total,
            Comments = GetS("comments"),
            MondayStart = GetS("mondayStart"),
            MondayEnd = GetS("mondayEnd"),
            TuesdayStart = GetS("tuesdayStart"),
            TuesdayEnd = GetS("tuesdayEnd"),
            WednesdayStart = GetS("wednesdayStart"),
            WednesdayEnd = GetS("wednesdayEnd"),
            ThursdayStart = GetS("thursdayStart"),
            ThursdayEnd = GetS("thursdayEnd"),
            FridayStart = GetS("fridayStart"),
            FridayEnd = GetS("fridayEnd"),
            SaturdayStart = GetS("saturdayStart"),
            SaturdayEnd = GetS("saturdayEnd"),
            SundayStart = GetS("sundayStart"),
            SundayEnd = GetS("sundayEnd"),
            CreatedAt = now,
        };
    }

    private static Dictionary<string, object?> TimesheetToDict(Timesheet ts) => new()
    {
        ["id"] = ts.Id,
        ["employeeId"] = ts.EmployeeId,
        ["weekStartDate"] = ts.WeekStartDate,
        ["weekEndDate"] = ts.WeekEndDate,
        ["status"] = ts.Status.ToString(),
        ["totalHours"] = ts.TotalHours,
        ["submittedAt"] = ts.SubmittedAt,
        ["approvedAt"] = ts.ApprovedAt,
        ["approvedBy"] = ts.ApprovedBy,
        ["rejectionComment"] = ts.RejectionComment,
        ["createdAt"] = ts.CreatedAt,
        ["updatedAt"] = ts.UpdatedAt,
    };

    private static Dictionary<string, object?> ReadTimesheetRow(NpgsqlDataReader reader)
    {
        object? Get(string col)
        {
            try { return reader[col] is DBNull ? null : reader[col]; } catch { return null; }
        }
        return new()
        {
            ["id"] = Get("id"),
            ["employeeId"] = Get("employee_id"),
            ["weekStartDate"] = Get("week_start_date"),
            ["weekEndDate"] = Get("week_end_date"),
            ["status"] = Get("status")?.ToString(),
            ["totalHours"] = Get("total_hours"),
            ["submittedAt"] = Get("submitted_at"),
            ["approvedAt"] = Get("approved_at"),
            ["approvedBy"] = Get("approved_by"),
            ["rejectionComment"] = Get("rejection_comment"),
            ["createdAt"] = Get("created_at"),
            ["updatedAt"] = Get("updated_at"),
        };
    }
}
